.github/workflows/upgrade.yml

```diff
@@ -56,22 +56,84 @@ jobs:
       - name: Set up Ginkgo CLI
         run: |
           go install github.com/onsi/ginkgo/v2/ginkgo@v2.19.1
+      
+      - name: Travel back in time to the before upgrade version
+        run: |
+          GIT_TAG="${{ github.event.inputs.beforeTagOrCommit }}"
+          PREVIOUS_BRANCH=$(git branch --show-current)
+          PREVIOUS_COMMIT=$(git rev-parse HEAD)
+          echo "Current at branch $PREVIOUS_BRANCH, commit $PREVIOUS_COMMIT."
 
-      - name: Prepare the fleet
-        run: cd test/upgrade && GIT_TAG="${{ github.event.inputs.beforeTagOrCommit }}" chmod +x setup.sh && ./setup.sh 3 && cd -
+          echo "PREVIOUS_COMMIT=$PREVIOUS_COMMIT" >> "$GITHUB_ENV"
+
+          if [ -z "${GIT_TAG}" ]
+          then
+              echo "No tag is specified; go back to the state tracked by the last known tag."
+              echo "Fetch all tags..."
+
+              git fetch --all
+              GIT_TAG=$(git describe --tags $(git rev-list --tags --max-count=1))
+              
+          else
+              echo "A tag is specified; go back to the state tracked by the specified tag."
+              echo "Fetch all tags..."
+
+              git fetch --all
+          fi
+
+          git checkout $GIT_TAG
+          echo "Checked out source code at $GIT_TAG."
+
+      - name: Prepare the fleet using the before upgrade version
+        run: cd test/upgrade && chmod +x setup.sh && ./setup.sh 3 && cd -
         env:
           KUBECONFIG: '/home/runner/.kube/config'
           HUB_SERVER_URL: 'https://172.19.0.2:6443'
+      
+      - name: Travel to the current state
+        # Note: Fleet always uses the version compatibility test suite from the 
+        # baseline commit, i.e., the commit that triggers the workflow.
+        run: |
+          echo "Returning to the current state..."
+          git checkout $PREVIOUS_COMMIT
```

Comment:
why is this not current commit?

Reply:
Hi Ryan! The upgrade pipeline often runs on dangling commits (commits that have not been checked as the PR hasn't been merged) so we couldn't just jump to the HEAD.